docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+have this only available in certain situations, for example when a
+`load_groups`, or similar, argument is passed to the templater.
+
+When this feature is implemented the `hosts` file will be able to be rendered
+in full in a single pass; since the information used in the rendering can
+change with any `configure` command it seems sensible to automatically
+re-render the `hosts` file after every `configure` command, similarly to the
+plan for `genders` after any `configure group` command. If we add this it will
+largely make the `hosts` command obsolete, although we may want to retain it
+for re-rendering after manually changing the template or a config file.
+
+We will also need to consider if we this how a `hosts` template other than the
+default will be selected for automatic re-rendering, or alternatively if we
+still need the ability to have multiple `hosts` templates within a single repo
+now - it should now be possible to have the `hosts` template handle any
+situations where we want to treat groups or nodes differently.
+
+
+## Command ordering
+
+With this new functionality certain commands will now require that other
+commands have been run first:
+
+- All `metal configure` commands will require the repo be set up (`repo use`
+  has been run) before running.
+
+- `metal build` will require that `metal configure all` has been run first,
+  i.e. that `/var/lib/metalware/config/all.yaml` has been generated, so any
+  necessary config has been set up before templating.
+
+- Additionally, `metal build -g $GROUP` will require that `metal configure
+  group $GROUP` has been run first, i.e. that
+  `/var/lib/metalware/config/groups/$GROUP.yaml` has been generated.
+
+- Note that a node config is not required, so a
+  `/var/lib/metalware/config/nodes/$NODE.yaml` file is not required to build
+  `$NODE`.
+
+
+## Templating
+
+When loading `alces.config` values from files in `/var/lib/metalware/config/`
+for use in templates, the same procedure will be followed as for loading the
+repo configs, i.e.
+
+- first `all.yaml` will be loaded;
+
+- if no `nodename` is passed, then nothing else will be loaded;
+
+- if a `nodename` is passed, all/any gender groups will be found for the node,
+  and any `groups/$GROUP.yaml` configs will be loaded for these in order of
+  precedence;
+
+- any `nodes/$NODE.yaml` file for the specific node will be loaded.
+
+When templating, attempting to use an `alces.config` value which is not set in
+the config should be a fatal error. All config values should be set up, as even
+if a particular config question is optional and nothing has been entered a
+value of `nil` should still be saved, in which case `nil` will be used when
+templating - however if the specified identifier is not present at all then
+this indicates the repo has an inconsistent `config` directory and
+`configure.yaml` file, which should be addressed.
+
+
+## Questions
+
+- May want to change `all` to `cluster` everywhere?
```

Comment:
I'd originally chosen all to try and avoid people thinking about platforms metalware sets up as single 'clusters', it can currently build an environment that would run several flight clusters for example. Appreciate 'all' might be a little too generic, maybe domain.yaml is a better word than cluster?